diff -ruN mlpack-3.1.1.orig/CMakeLists.txt mlpack-3.1.1/CMakeLists.txt
--- mlpack-3.1.1.orig/CMakeLists.txt	2019-05-26 20:18:49.000000000 -0600
+++ mlpack-3.1.1/CMakeLists.txt	2019-09-07 14:11:07.714156105 -0600
@@ -316,15 +316,17 @@
 find_package(Ensmallen 1.10.0)
 if (NOT ENSMALLEN_FOUND)
   if (DOWNLOAD_ENSMALLEN)
-    file(DOWNLOAD http://www.ensmallen.org/files/ensmallen-latest.tar.gz
-        "${CMAKE_BINARY_DIR}/deps/ensmallen-latest.tar.gz"
+    file(DOWNLOAD http://www.ensmallen.org/files/ensmallen-1.16.2.tar.gz
+        "${CMAKE_BINARY_DIR}/deps/ensmallen-1.16.2.tar.gz"
         STATUS ENS_DOWNLOAD_STATUS_LIST LOG ENS_DOWBLOAD_LOG
         SHOW_PROGRESS)
     list(GET ENS_DOWNLOAD_STATUS_LIST 0 ENS_DOWNLOAD_STATUS)
     if (ENS_DOWNLOAD_STATUS EQUAL 0)
       execute_process(COMMAND ${CMAKE_COMMAND} -E
-          tar xzf "${CMAKE_BINARY_DIR}/deps/ensmallen-latest.tar.gz"
+          tar xzf "${CMAKE_BINARY_DIR}/deps/ensmallen-1.16.2.tar.gz"
           WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/deps/")
+      execute_process(COMMAND
+          rm "${CMAKE_BINARY_DIR}/deps/ensmallen-1.16.2.tar.gz")
 
       # Get the name of the directory.
       file (GLOB ENS_DIRECTORIES RELATIVE "${CMAKE_BINARY_DIR}/deps/"
diff -ruN mlpack-3.1.1.orig/src/mlpack/methods/cf/decomposition_policies/nmf_method.hpp mlpack-3.1.1/src/mlpack/methods/cf/decomposition_policies/nmf_method.hpp
--- mlpack-3.1.1.orig/src/mlpack/methods/cf/decomposition_policies/nmf_method.hpp	2019-05-26 20:18:49.000000000 -0600
+++ mlpack-3.1.1/src/mlpack/methods/cf/decomposition_policies/nmf_method.hpp	2019-09-07 14:03:05.074679944 -0600
@@ -129,7 +129,7 @@
     // matrix with the Mahalanobis distance where M^{-1} = W^T W.  So, we'll
     // decompose M^{-1} = L L^T (the Cholesky decomposition), and then multiply
     // H by L^T. Then we can perform nearest neighbor search.
-    arma::mat l = arma::chol(w.t() * w);
+    arma::mat l = arma::chol(w.t() * w + 1e-20 * arma::eye<arma::mat>(w.n_cols, w.n_cols));
     arma::mat stretchedH = l * h; // Due to the Armadillo API, l is L^T.
 
     // Temporarily store feature vector of queried users.
