diff -ruN mlpack-3.1.1/src/mlpack/methods/cf/decomposition_policies/nmf_method.hpp mlpack-3.1.1.new/src/mlpack/methods/cf/decomposition_policies/nmf_method.hpp
--- mlpack-3.1.1/src/mlpack/methods/cf/decomposition_policies/nmf_method.hpp	2019-05-26 20:18:49.000000000 -0600
+++ mlpack-3.1.1.new/src/mlpack/methods/cf/decomposition_policies/nmf_method.hpp	2019-09-04 11:05:50.902333204 -0600
@@ -129,7 +129,7 @@
     // matrix with the Mahalanobis distance where M^{-1} = W^T W.  So, we'll
     // decompose M^{-1} = L L^T (the Cholesky decomposition), and then multiply
     // H by L^T. Then we can perform nearest neighbor search.
-    arma::mat l = arma::chol(w.t() * w);
+    arma::mat l = arma::chol(w.t() * w + 1e-20 * arma::eye<arma::mat>(w.n_cols, w.n_cols));
     arma::mat stretchedH = l * h; // Due to the Armadillo API, l is L^T.
 
     // Temporarily store feature vector of queried users.
